File: packages/modules/Bluetooth/system/gd/hci/facade/controller_facade.cc
/*
 * Copyright 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "hci/facade/controller_facade.h"

#include <condition_variable>
#include <memory>
#include <mutex>

#include "blueberry/facade/hci/controller_facade.grpc.pb.h"
#include "blueberry/facade/hci/controller_facade.pb.h"
#include "common/bind.h"
#include "common/blocking_queue.h"
#include "grpc/grpc_event_queue.h"
#include "hci/address.h"
#include "hci/controller.h"

using ::grpc::ServerAsyncResponseWriter;
using ::grpc::ServerAsyncWriter;
using ::grpc::ServerContext;

namespace bluetooth {
namespace hci {
namespace facade {

using namespace blueberry::facade::hci;
using blueberry::facade::BluetoothAddress;

class ControllerFacadeService : public ControllerFacade::Service {
 public:
  ControllerFacadeService(Controller* controller, ::bluetooth::os::Handler*) : controller_(controller) {}

  ::grpc::Status GetMacAddress(
      ::grpc::ServerContext* context, const ::google::protobuf::Empty* request, BluetoothAddress* response) override {
    Address local_address = controller_->GetMacAddress();
    response->set_address(local_address.ToString());
    return ::grpc::Status::OK;
  }

  ::grpc::Status GetLocalName(::grpc::ServerContext* context, const ::google::protobuf::Empty* request,
                              NameMsg* response) override {
    std::string local_name = controller_->GetLocalName();
    response->set_name(local_name);
    return ::grpc::Status::OK;
  }

  ::grpc::Status WriteLocalName(::grpc::ServerContext* context, const NameMsg* request,
                                ::google::protobuf::Empty* response) override {
    controller_->WriteLocalName(request->name());
    return ::grpc::Status::OK;
  }

  ::grpc::Status IsSupportedCommand(
      ::grpc::ServerContext* context, const OpCodeMsg* request, SupportedMsg* response) override {
    bool ret = controller_->IsSupported(static_cast<OpCode>(request->op_code()));
    response->set_supported(ret);
    return ::grpc::Status::OK;
  }

#define SUPPORTED_API(name)                                                                                        \
  ::grpc::Status name(                                                                                             \
      ::grpc::ServerContext* context, const ::google::protobuf::Empty* request, SupportedMsg* response) override { \
    response->set_supported(controller_->name());                                                                  \
    return ::grpc::Status::OK;                                                                                     \
  }

  SUPPORTED_API(SupportsSimplePairing)
  SUPPORTED_API(SupportsSecureConnections)
  SUPPORTED_API(SupportsSimultaneousLeBrEdr)
  SUPPORTED_API(SupportsInterlacedInquiryScan)
  SUPPORTED_API(SupportsRssiWithInquiryResults)
  SUPPORTED_API(SupportsExtendedInquiryResponse)
  SUPPORTED_API(SupportsRoleSwitch)
  SUPPORTED_API(Supports3SlotPackets)
  SUPPORTED_API(Supports5SlotPackets)
  SUPPORTED_API(SupportsClassic2mPhy)
  SUPPORTED_API(SupportsClassic3mPhy)
  SUPPORTED_API(Supports3SlotEdrPackets)
  SUPPORTED_API(Supports5SlotEdrPackets)
  SUPPORTED_API(SupportsSco)
  SUPPORTED_API(SupportsHv2Packets)
  SUPPORTED_API(SupportsHv3Packets)
  SUPPORTED_API(SupportsEv3Packets)
  SUPPORTED_API(SupportsEv4Packets)
  SUPPORTED_API(SupportsEv5Packets)
  SUPPORTED_API(SupportsEsco2mPhy)
  SUPPORTED_API(SupportsEsco3mPhy)
  SUPPORTED_API(Supports3SlotEscoEdrPackets)
  SUPPORTED_API(SupportsHoldMode)
  SUPPORTED_API(SupportsSniffMode)
  SUPPORTED_API(SupportsParkMode)
  SUPPORTED_API(SupportsNonFlushablePb)
  SUPPORTED_API(SupportsSniffSubrating)
  SUPPORTED_API(SupportsEncryptionPause)
  SUPPORTED_API(SupportsBle)
  SUPPORTED_API(SupportsBleEncryption)
  SUPPORTED_API(SupportsBleConnectionParametersRequest)
  SUPPORTED_API(SupportsBleExtendedReject)
  SUPPORTED_API(SupportsBlePeripheralInitiatedFeaturesExchange)
  SUPPORTED_API(SupportsBlePing)
  SUPPORTED_API(SupportsBleDataPacketLengthExtension)
  SUPPORTED_API(SupportsBlePrivacy)
  SUPPORTED_API(SupportsBleExtendedScannerFilterPolicies)
  SUPPORTED_API(SupportsBle2mPhy)
  SUPPORTED_API(SupportsBleStableModulationIndexTx)
  SUPPORTED_API(SupportsBleStableModulationIndexRx)
  SUPPORTED_API(SupportsBleCodedPhy)
  SUPPORTED_API(SupportsBleExtendedAdvertising)
  SUPPORTED_API(SupportsBlePeriodicAdvertising)
  SUPPORTED_API(SupportsBleChannelSelectionAlgorithm2)
  SUPPORTED_API(SupportsBlePowerClass1)
  SUPPORTED_API(SupportsBleMinimumUsedChannels)
  SUPPORTED_API(SupportsBleConnectionCteRequest)
  SUPPORTED_API(SupportsBleConnectionCteResponse)
  SUPPORTED_API(SupportsBleConnectionlessCteTransmitter)
  SUPPORTED_API(SupportsBleConnectionlessCteReceiver)
  SUPPORTED_API(SupportsBleAntennaSwitchingDuringCteTx)
  SUPPORTED_API(SupportsBleAntennaSwitchingDuringCteRx)
  SUPPORTED_API(SupportsBleReceivingConstantToneExtensions)
  SUPPORTED_API(SupportsBlePeriodicAdvertisingSyncTransferSender)
  SUPPORTED_API(SupportsBlePeriodicAdvertisingSyncTransferRecipient)
  SUPPORTED_API(SupportsBleSleepClockAccuracyUpdates)
  SUPPORTED_API(SupportsBleRemotePublicKeyValidation)
  SUPPORTED_API(SupportsBleConnectedIsochronousStreamCentral)
  SUPPORTED_API(SupportsBleConnectedIsochronousStreamPeripheral)
  SUPPORTED_API(SupportsBleIsochronousBroadcaster)
  SUPPORTED_API(SupportsBleSynchronizedReceiver)
  SUPPORTED_API(SupportsBleIsochronousChannelsHostSupport)
  SUPPORTED_API(SupportsBlePowerControlRequest)
  SUPPORTED_API(SupportsBlePowerChangeIndication)
  SUPPORTED_API(SupportsBlePathLossMonitoring)

  ::grpc::Status GetLeNumberOfSupportedAdvertisingSets(
      ::grpc::ServerContext* context, const ::google::protobuf::Empty* request, SingleValueMsg* response) override {
    uint8_t ret = controller_->GetLeNumberOfSupportedAdverisingSets();
    response->set_value(ret);
    return ::grpc::Status::OK;
  }

 private:
  Controller* controller_;
};

void ControllerFacadeModule::ListDependencies(ModuleList* list) const {
  ::bluetooth::grpc::GrpcFacadeModule::ListDependencies(list);
  list->add<Controller>();
}

void ControllerFacadeModule::Start() {
  ::bluetooth::grpc::GrpcFacadeModule::Start();
  service_ = new ControllerFacadeService(GetDependency<Controller>(), GetHandler());
}

void ControllerFacadeModule::Stop() {
  delete service_;
  ::bluetooth::grpc::GrpcFacadeModule::Stop();
}

::grpc::Service* ControllerFacadeModule::GetService() const {
  return service_;
}

const ModuleFactory ControllerFacadeModule::Factory =
    ::bluetooth::ModuleFactory([]() { return new ControllerFacadeModule(); });

}  // namespace facade
}  // namespace hci
}  // namespace bluetooth

File: packages/modules/Bluetooth/system/gd/rust/stack/src/hci/controller_facade.rs
//! Controller facade

use crate::hci::controller::{null_terminated_to_string, ControllerExports};
use crate::hci::Hci;
use bt_common::GrpcFacade;
use bt_facade_proto::common::BluetoothAddress;
use bt_facade_proto::controller_facade::{NameMsg, OpCodeMsg, SingleValueMsg, SupportedMsg};
use bt_facade_proto::controller_facade_grpc::{create_controller_facade, ControllerFacade};
use bt_facade_proto::empty::Empty;
use bt_packets::hci::{OpCode, ReadLocalNameBuilder, WriteLocalNameBuilder};
use gddi::{module, provides, Stoppable};
use grpcio::*;
use num_traits::FromPrimitive;
use std::sync::Arc;

module! {
    controller_facade_module,
    providers {
        ControllerFacadeService => provide_facade,
    }
}

#[provides]
async fn provide_facade(exports: Arc<ControllerExports>, hci: Hci) -> ControllerFacadeService {
    ControllerFacadeService { exports, hci }
}

/// Controller facade service
#[allow(missing_docs)]
#[derive(Clone, Stoppable)]
pub struct ControllerFacadeService {
    pub exports: Arc<ControllerExports>,
    hci: Hci,
}

impl GrpcFacade for ControllerFacadeService {
    fn into_grpc(self) -> grpcio::Service {
        create_controller_facade(self)
    }
}

impl ControllerFacade for ControllerFacadeService {
    fn get_mac_address(
        &mut self,
        ctx: RpcContext<'_>,
        _req: Empty,
        sink: UnarySink<BluetoothAddress>,
    ) {
        let clone = self.clone();
        ctx.spawn(async move {
            let mut address = BluetoothAddress::new();
            address.set_address(clone.exports.address.bytes.to_vec());
            sink.success(address).await.unwrap();
        });
    }

    fn write_local_name(&mut self, ctx: RpcContext<'_>, req: NameMsg, sink: UnarySink<Empty>) {
        let mut clone = self.clone();
        let mut builder = WriteLocalNameBuilder { local_name: [0; 248] };
        builder.local_name[0..req.get_name().len()].copy_from_slice(req.get_name());
        ctx.spawn(async move {
            clone.hci.commands.send(builder.build()).await;
            sink.success(Empty::default()).await.unwrap();
        });
    }

    fn get_local_name(&mut self, ctx: RpcContext<'_>, _req: Empty, sink: UnarySink<NameMsg>) {
        let mut clone = self.clone();
        ctx.spawn(async move {
            let local_name = null_terminated_to_string(
                clone.hci.commands.send(ReadLocalNameBuilder {}).await.get_local_name(),
            )
            .into_bytes();
            let mut msg = NameMsg::new();
            msg.set_name(local_name);
            sink.success(msg).await.unwrap();
        });
    }

    fn is_supported_command(
        &mut self,
        ctx: RpcContext<'_>,
        op_code_msg: OpCodeMsg,
        sink: UnarySink<SupportedMsg>,
    ) {
        let clone = self.clone();
        let opcode = OpCode::from_u32(op_code_msg.get_op_code()).unwrap();
        ctx.spawn(async move {
            let mut supported_msg = SupportedMsg::new();
            supported_msg.set_supported(clone.exports.commands.is_supported(opcode));
            sink.success(supported_msg).await.unwrap();
        });
    }

    fn get_le_number_of_supported_advertising_sets(
        &mut self,
        ctx: RpcContext<'_>,
        _: Empty,
        sink: UnarySink<SingleValueMsg>,
    ) {
        let clone = self.clone();
        ctx.spawn(async move {
            let mut msg = SingleValueMsg::new();
            msg.set_value(clone.exports.le_supported_advertising_sets.into());
            sink.success(msg).await.unwrap();
        });
    }

    fn supports_simple_pairing(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }

    fn supports_secure_connections(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }

    fn supports_simultaneous_le_br_edr(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }

    fn supports_interlaced_inquiry_scan(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }

    fn supports_rssi_with_inquiry_results(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }

    fn supports_extended_inquiry_response(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }

    fn supports_role_switch(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }

    fn supports3_slot_packets(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }

    fn supports5_slot_packets(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }

    fn supports_classic2m_phy(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }

    fn supports_classic3m_phy(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }

    fn supports3_slot_edr_packets(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }

    fn supports5_slot_edr_packets(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }

    fn supports_sco(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }

    fn supports_hv2_packets(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_hv3_packets(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_ev3_packets(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_ev4_packets(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }

    fn supports_ev5_packets(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_esco2m_phy(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_esco3m_phy(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports3_slot_esco_edr_packets(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_hold_mode(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_sniff_mode(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_park_mode(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_non_flushable_pb(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_sniff_subrating(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_encryption_pause(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_ble_encryption(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_ble_connection_parameters_request(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_extended_reject(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_peripheral_initiated_features_exchange(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_ping(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_ble_data_packet_length_extension(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_privacy(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_ble_extended_scanner_filter_policies(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble2m_phy(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_ble_stable_modulation_index_tx(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_stable_modulation_index_rx(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_coded_phy(&mut self, _: RpcContext<'_>, _: Empty, _: UnarySink<SupportedMsg>) {
        todo!()
    }
    fn supports_ble_extended_advertising(
        &mut self,
        ctx: RpcContext<'_>,
        _: Empty,
        sink: UnarySink<SupportedMsg>,
    ) {
        let clone = self.clone();
        ctx.spawn(async move {
            let mut supported_msg = SupportedMsg::new();
            supported_msg.set_supported(clone.exports.le_features.extended_advertising);
            sink.success(supported_msg).await.unwrap();
        });
    }
    fn supports_ble_periodic_advertising(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_channel_selection_algorithm2(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_power_class1(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_minimum_used_channels(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_connection_cte_request(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_connection_cte_response(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_connectionless_cte_transmitter(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_connectionless_cte_receiver(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_antenna_switching_during_cte_tx(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_antenna_switching_during_cte_rx(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_receiving_constant_tone_extensions(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_periodic_advertising_sync_transfer_sender(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_periodic_advertising_sync_transfer_recipient(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_sleep_clock_accuracy_updates(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_remote_public_key_validation(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_connected_isochronous_stream_central(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_connected_isochronous_stream_peripheral(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_isochronous_broadcaster(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_synchronized_receiver(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_isochronous_channels_host_support(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_power_control_request(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_power_change_indication(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
    fn supports_ble_path_loss_monitoring(
        &mut self,
        _: RpcContext<'_>,
        _: Empty,
        _: UnarySink<SupportedMsg>,
    ) {
        todo!()
    }
}
